Rewrite 0.5 `@boundscheck` to a no-op in 0.4

c42f opened the pull request:
Fixes #182.

`@boundscheck` existed in 0.4, but had a very different meaning, being used internally by `@inbounds`.  There's no equivalent to the 0.5 meaning, so rewrite it to a no-op in `@compat`.

I returned `nothing` for consistency since the new `@boundscheck` has an `Expr(:boundscheck, :pop)` as the last statement in the block.  I'd be happy to hear if there's a more appropriate way to do this.

yuyichao commented:
Please quote your code...

yuyichao commented:
Also as noted in https://github.com/JuliaLang/Compat.jl/issues/182#issuecomment-223726410 this should use `Compat.@boundscheck` instead to not break code on 0.4.

c42f commented:
I did quote it, just had a typo :-/ Apologies to anyone who got `@mentioned`

c42f commented:
Ok, I now check the number of arguments to `@boundscheck` for compatibility.  Too bad that `@boundscheck` was even exported in 0.4.

Still needs some tests I suppose, and I guess I should add the `@propagate_inbounds` macro as part of this too.

c42f commented:
Hmm.  `@propagate_inbounds` isn't exported from Base, but it's in the public docs.  Bug in Base?  Maybe I should leave that alone for now.

c42f commented:
Tests and docs added, should be good to go pending CI.

I punted on `@propagate_inbounds` for now.

yuyichao commented:
Try to avoid using `@` in commit messages too. (And as you see, quoting doesn't work there...)

yuyichao commented:
LGTM.

c42f commented:
Argh :-(  Luckily `@compat` doesn't seem to be an active github user.

Thanks for the review.

c42f commented:
Clarification about propagate_inbounds requested see JuliaLang/julia#16752

c42f commented:
Ugh, so I wonder how I'm meant to test this!  The latest reason for the test failure in 0.5 is because `Pkg.test()` is called with `julia --check-bounds=yes`, which turns off the action of `@inbounds`... which is precisely what I'm attempting to test here.

I commented that part of the test out - I'm not sure how else to keep this simple.

yuyichao commented:
You can just delete the 0.5 test. It's purely testing a `Base` function anyway... You usually only need to worry about the shared function provided by Compat.

c42f commented:
I'll kill it off then, might have been overkill.  The point about testing 0.5 functionality was to test that rewriting the syntax didn't mess with the new version.

c42f commented:
Eh, that was rather a lot of drama for a simple change, hopefully CI will pass now!

c42f commented:
Can we merge this?  It's been ready for a while and I'd say in its current state is pretty simple and uncontroversial.

TotalVerb commented:
Can you rebase the commit? There are conflicts preventing merging.

c42f commented:
I rebased and fixed the conflicts.

There's some appveyor failures here so far, but that's just complaining about string handling in nightly which seems unrelated.

c42f commented:
Ok, Travis looks good.  I think this is good to go again.

TotalVerb commented:
Agreed that Appveyor looks spurious.
